tuxedo-parachain-core/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+
+/// An abstraction over reading the ambiently available relay parent number.
+/// This allows it to be mocked during tests and require actual externalities.
+pub trait GetRelayParentNumberStorage {
+    fn get() -> u32;
+}
+
+/// An abstraction over setting the ambiently available relay parent number.
+/// This allows it to be mocked during tests and require actual externalities.
+pub trait SetRelayParentNumberStorage {
+    fn set(new_parent_number: u32);
+}
+
+/// A public interface for accessing and mutating the relay parent number. This is
+/// expected to be called from the parachain piece
+pub enum RelayParentNumberStorage {}
+
+impl GetRelayParentNumberStorage for RelayParentNumberStorage {
+    fn get() -> u32 {
+        let encoded = sp_io::storage::get(RELAY_PARENT_NUMBER_KEY)
+            .expect("Some relay parent number should always be stored");
+        Decode::decode(&mut &encoded[..])
+            .expect("properly encoded relay parent number should have been stored.")
+    }
+}
+
+impl SetRelayParentNumberStorage for RelayParentNumberStorage {
+    fn set(new_parent_number: u32) {
+        sp_io::storage::set(RELAY_PARENT_NUMBER_KEY, &new_parent_number.encode());
+    }
+}
+
+/// A mock version of the RelayParentNumberStorage that can be used in tests without externalities.
+pub enum MockRelayParentNumberStorage {}
+
+impl SetRelayParentNumberStorage for MockRelayParentNumberStorage {
+    fn set(_new_parent_number: u32) {}
+}
+
+/// Basically the same as
+/// [`ValidationParams`](polkadot_parachain_primitives::primitives::ValidationParams), but a little
+/// bit optimized for our use case here.
+///
+/// `block_data` and `head_data` are represented as [`bytes::Bytes`] to make them reuse
+/// the memory of the input parameter of the exported `validate_blocks` function.
+///
+/// The layout of this type must match exactly the layout of
+/// [`ValidationParams`](polkadot_parachain_primitives::primitives::ValidationParams) to have the
+/// same SCALE encoding.
+#[derive(parity_scale_codec::Decode)]
+#[cfg_attr(feature = "std", derive(parity_scale_codec::Encode))]
+#[doc(hidden)]
+pub struct MemoryOptimizedValidationParams {
+    pub parent_head: bytes::Bytes,
+    pub block_data: bytes::Bytes,
+    pub relay_parent_number: cumulus_primitives_core::relay_chain::BlockNumber,
+    pub relay_parent_storage_root: cumulus_primitives_core::relay_chain::Hash,
+}
+
+/// Register the `validate_block` function that is used by parachains to validate blocks on a
+/// validator.
+///
+/// Does *nothing* when `std` feature is enabled.
+///
+/// Expects as parameters the Block type, the OuterVerifier, and the OuterConstraintChecker.
+pub use tuxedo_register_validate_block::register_validate_block;
+
+/// A wrapper type around Cumulus's ParachainInherentData ype that can be stored.
+/// Having to do this wrapping is one more reason to abandon this UtxoData trait,
+/// and go for a more strongly typed aggregate type approach.
```

Comment:
Are you sure you want to leave this comment in the docs? 
I think it would be better to link it to an issue to discuss the removal of the UtxoData trait.

Reply:
Addressed in e163f0362